File: pset6/vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

int get_num_value(char c);

int main(int argc, string argv[])
{
    if (argc == 2) //check that there is exactly 1 argument
    {
        for (int i = 0, len = strlen(argv[1]); i < len; i++) //iterate through each character in the argument
        {
            if (isalpha(argv[1][i]) == false) //if any are not digits, return an error
            {
                printf("Usage: ./vigenere keyword\n");
                return 1;
            }
        }
    }
    else //if there is more than 1 argument return an error
    {
        printf("Usage: ./vigenere keyword\n");
        return 1;
    }

    string s_key = argv[1]; //make the key an int
    int keylen = strlen(s_key);
    
    string ptext = get_string("plaintext: "); //ask the user for the plaintext

    printf("ciphertext: ");
    int j = 0;
    for (int i = 0, len = strlen(ptext); i < len; i++) //iterate through each character in ptext
    {
        if (j >= keylen) { j = 0; }
        
        int i_key;
        i_key = get_num_value(s_key[j]);
        //printf("i_key:%i j:%i ptext:%c    ", i_key, j, ptext[i]);

        char ctext;
        if (isupper(ptext[i]))
        {
            ctext = ((get_num_value(ptext[i]) + i_key) % 26) + 65;
            printf("%c", ctext);
            j++;
        }
        else if (islower(ptext[i]))
        {
            ctext = ((get_num_value(ptext[i]) + i_key) % 26) + 97;
            printf("%c", ctext);
            j++;
        }
        else //if it's not a letter, print as it is
        {
            printf("%c", ptext[i]);
        }
    }
    printf("\n");
}

int get_num_value(char c) //calculate the numerical value of each letter (1-26)
{
    int v;
    if (isupper(c))
    {
        v = (c - 65);
    }
    else
    {
        v = (c - 97);
    }
    return v;
}

File: pset6/vigenere.py
from sys import argv, exit

UPPERCASE = 65
LOWERCASE = 97

def main():
    if len(argv) != 2:
        print('Usage: python vigenere.py key')
        exit(1)
    else:
        if not argv[1].isalpha():
            print('Usage: python vigenere.py key')
            exit(1)


    k_word = argv[1]
    k_len = len(k_word)
    p_text = input('plaintext: ')
    c_text = ""

    j = 0

    for i in p_text:
        p_val, is_up = get_num_value(i)
        k_val, _k_is_up = get_num_value(k_word[j])

        out = p_val + k_val
        if out > 25:
            out -= 26

        if is_up and p_val >= 0:
            out += UPPERCASE
        elif p_val >= 0:
            out += LOWERCASE
        else:
            out = ord(i)
        c_text += chr(out)

        if p_val >= 0:
           j += 1

        if j == k_len:
           j = 0

    print("ciphertext: " + c_text)


def get_num_value(i):
    if i.isupper():
        val = ord(i) - UPPERCASE
        return val, True
    elif i.islower():
        val = ord(i) - LOWERCASE
        return val, False
    else:
        return -1, False


if __name__ == '__main__':
    main()
